composer/trainer/trainer.py

```diff
@@ -180,7 +180,65 @@ def _get_initial_grad_accum(grad_accum: Union[int, str]):
 
 def _is_cuda_oom(e: RuntimeError):
     """Determines if error is CUDA Out of Memory and if adaptive_grad_accum is enabled."""
-    return 'CUDA out of memory' in str(e)
+    if 'CUDA out of memory' in str(e):
+        return True
+    # With batch_norm, large batch sizes sometimes result in cuDNN instead of Cuda OOMs.
+    if 'cuDNN error: CUDNN_STATUS_NOT_SUPPORTED. This error may appear if you passed in a non-contiguous input.' in str(
+            e):
+        warnings.warn(
+            f'Encountered "cuDNN error: CUDNN_STATUS_NOT_SUPPORTED. This error may appear if you passed in a non-contiguous input." This can happen when the batch_size is too large for the GPU so auto grad_accum will rerun with a higher grad_accum value, but there may be a user error with non-contiguous inputs.'
+        )
+        return True
+    return False
+
+
+def _adjust_grad_accum(state: State, device_batch_size):
+    """Adjust grad_accum if we encounter OOM.
+
+    Args:
+        state (State): State of trainer.
+        device_batch_size (int): Batch size.
+    """
+    # If any rank hit CUDA OOM, update grad_accum and retry. Raise runtime error if training 1 sample
+    # at a time still resulted in CUDA out of memory.
+    if state.grad_accum == device_batch_size:
+        raise RuntimeError(('CUDA out of memory. The train loop failed with an internal microbatch of size 1.'
+                            'The GPU does not have enough memory to process even 1 sample.'))
+    else:
+        original_grad_accum = state.grad_accum
+        state.grad_accum = min(2 * state.grad_accum, device_batch_size)
+        warnings.warn(
+            RuntimeWarning('CUDA out of memory detected. Gradient Accumulation, the number of train microbatches, '
+                           f'increased from {original_grad_accum} -> {state.grad_accum}, '
+                           'and the batch will be retrained with a '
+                           f'micro-batchsize of {device_batch_size // state.grad_accum}'))
+    # Clear gradients in case failure happened during backwards pass
+    for optimizer in state.optimizers:
+        optimizer.zero_grad()
```

Comment:
@dskhudia  this should resolve the += gradient issue, correct? Not as familiar with internals so I'd like to verify

Reply:
This should do it.